src/Microsoft.Identity.Client/TokenCache.cs

```diff
@@ -204,1024 +204,1046 @@ internal void OnBeforeWrite(TokenCacheNotificationArgs args)
+                OnAfterAccess(args);
 
-            // this is OBO flow. match the cache entry with assertion hash,
-            // Authority, ScopeSet and client Id.
-            if (requestParams.UserAssertion != null)
-            {
-                requestParams.RequestContext.Logger.Info("Filtering by user assertion...");
-                tokenCacheItems =
-                    tokenCacheItems.Where(
-                            item =>
-                                !string.IsNullOrEmpty(item.UserAssertionHash) &&
-                                item.UserAssertionHash.Equals(requestParams.UserAssertion.AssertionHash, StringComparison.OrdinalIgnoreCase))
-                        .ToList();
-            }
-            else
-            {
-                if (!requestParams.IsClientCredentialRequest)
+                // this is OBO flow. match the cache entry with assertion hash,
+                // Authority, ScopeSet and client Id.
+                if (requestParams.UserAssertion != null)
                 {
-                    requestParams.RequestContext.Logger.Info("Filtering by user identifier...");
-                    // filter by identifier of the user instead
+                    requestParams.RequestContext.Logger.Info("Filtering by user assertion...");
                     tokenCacheItems =
-                        tokenCacheItems
-                            .Where(item => item.HomeAccountId.Equals(requestParams.Account?.HomeAccountId.Identifier, StringComparison.OrdinalIgnoreCase))
+                        tokenCacheItems.Where(
+                                item =>
+                                    !string.IsNullOrEmpty(item.UserAssertionHash) &&
+                                    item.UserAssertionHash.Equals(requestParams.UserAssertion.AssertionHash, StringComparison.OrdinalIgnoreCase))
                             .ToList();
                 }
-            }
-
-            // no match found after initial filtering
-            if (!tokenCacheItems.Any())
-            {
-                requestParams.RequestContext.Logger.Info("No matching entry found for user or assertion");
-                return null;
-            }
+                else
+                {
+                    if (!requestParams.IsClientCredentialRequest)
+                    {
+                        requestParams.RequestContext.Logger.Info("Filtering by user identifier...");
+                        // filter by identifier of the user instead
+                        tokenCacheItems =
+                            tokenCacheItems
+                                .Where(item => item.HomeAccountId.Equals(requestParams.Account?.HomeAccountId.Identifier, StringComparison.OrdinalIgnoreCase))
+                                .ToList();
+                    }
 
-            requestParams.RequestContext.Logger.Info("Matching entry count -" + tokenCacheItems.Count);
+                    // filter by authority in the auth request params
```

Comment:
>// filter by authority in the auth request params [](start = 20, length = 49)

imo, this can be removed. you have a method with a meaningful name. the method could have an xml comment if you prefer.